examples/plot_replay_cache.py

```diff
@@ -0,0 +1,143 @@
+
+iris = load_iris()
+X = iris.data
+y = iris.target
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=1)
+batch_size = 5
+n_iter = 10
+
+model = SVC(probability=True)
+
+samplers = [
+    ('margin', MarginSampler(model, batch_size))
+]
+
+#############################################################################
+# This function runs an experiment and optionally generate an error to test
+# the resuming capability.
+#
+# Note the use of the GrowingIndex utils that facilitate the handing of
+# indices in an active learning experiment.
+
+def test_run(with_error=False, compute_metric=None):
+
+    for sampler_name, sampler in samplers:
+
+        config = dict(sampler=sampler_name)
+
+        with ReplayCache('./cache', './cache.db', keys=config) as cache:
+
+            index = GrowingIndex(X_train.shape[0])
+
+            # Add at least one sample from each class
+            index.add_to_selected([np.where(y_train == i)[0][0] for i in np.unique(y)])
+
+            selected = cache.variable('selected', index.selected)
+
+            for j, prev_selected in cache.iter(range(n_iter), selected.previous()):
+                print('Computing iteration {}'.format(j))
+                index.resume(prev_selected)
+
+                model.fit(X_train[prev_selected], y_train[prev_selected])
+                sampler.fit(X_train[prev_selected], y_train[prev_selected])
+                index.add_to_selected(sampler.select_samples(X_train[index.non_selected]))
+                selected.set(index.selected)
+
+                if with_error and j == 5:
+                    raise ValueError('Simulated Error')
+
+            if compute_metric is not None:
+                cache.compute_metric('metric', compute_metric, selected.previous(), selected.current())
+
+
+#############################################################################
+# We run this function and force an error to happen. We then see how the
+# cache stores these values in a human readable way.
+#
+# We see that all selected indices have been kept up until the 4th iteration
+# (since an error happened at iteration 5).
+
+try:
+    test_run(with_error=True)
+except ValueError as e:
+    print('ValueError raised: ' + str(e))
+list_files('./cache')
+
+#############################################################################
+# We run the same function without error. In this case, we see that the 4
+# first iterations are skipped. The code is not even executed. Afterward,
+# the cache contains the data for all iterations.
+
+test_run()
+list_files('./cache')
+
+#############################################################################
+# Being a bit paranoid, we would like to check what cardinal does. For that,
+# we compute the batch size of each iteration. Fortunately, we have cached
+# the variable `selected` and therefore, we can replay the experiment.
+
+
+def calc_batch_size(previous, current):
```

Comment:
It's a bit complex and hard to read, it's only for paranoid people, not for a base example, and it adds a bit of complexity to the main function with the compute_metric thing. So I would trash all that.

Reply:
The problem is that the interest of the replay cache is to be able to compute a metric afterward. I could split it in two examples.

Reply:
Split in 2 examples